### PR TITLE
Read/Unread count update

### LINKED ISSUE #245: Remove unread/read count, just "new"

### LINKED ISSUE #278: Change category sort to 'title'

### PR DESCRIPTION
**Pull request by swglad**
Resolves #245. Closes #278.

### CODE REVIEW

File: sourcemash/api/categories.py [hunk 1]
```diff
@@ -1,11 +1,20 @@
...
-from sourcemash.models import Item, Category, UserItem
+from sourcemash.models import Item, Category, UserItem, UserCategory
 
 from sqlalchemy import func
 
+class isUnread(fields.Raw):
+    def output(self, key, category):
+        try:
+            unread = UserCategory.query.filter_by(user=current_user, category_id=category.id).one().unread
+        except:
```
Comment (alexgerstein) [inline sourcemash/api/categories.py]: Let's specify this except, like seen here: http://stackoverflow.com/questions/18110033/getting-first-row-from-sqlalchemy

File: sourcemash/api/categories.py [hunk 2]
```diff
@@ -13,18 +22,46 @@
...
 
 category_status_fields = {
-  'item_count': fields.Integer,
-  'unread_count': fields.Integer
+  'unread': isUnread
 }
 category_status_fields = dict(category_fields, **category_status_fields)
 
+category_status_count_fields = {
+  'item_count': fields.Integer,
```
Comment (alexgerstein) [inline sourcemash/api/categories.py]: Do we still want to bother with item count at all?

Comment (swglad) [inline sourcemash/api/categories.py]: We do it for feeds so is it weird not to do for categories? That being said, I’m not too attached and wouldn’t have any problem cutting it.

File: sourcemash/api/categories.py [hunk 3]
```diff
@@ -44,24 +81,16 @@ def get(self):
                              .all()
 
         for category, count in categories:
-            read_item_count = UserItem.query \
```
Comment (alexgerstein) [inline sourcemash/api/categories.py]: It's beautiful seeing this disappear...

File: sourcemash/api/feeds.py [hunk 4]
```diff
@@ -27,21 +27,17 @@ def output(self, key, feed):
...
-class getUnreadCount(fields.Raw):
+class isUnread(fields.Raw):
     def output(self, key, feed):
         if not current_user.is_authenticated():
-            return 0
+            return True
+
+        try:
+            unread = UserFeed.query.filter_by(user=current_user, feed_id=feed.id).one().unread
+        except:
```
Comment (alexgerstein) [inline sourcemash/api/feeds.py]: Specify except here too

File: sourcemash/frontend/static/js/views/items_view.js [hunk 5]
```diff
@@ -14,13 +15,21 @@ Sourcemash.Views.ItemsView = Backbone.View.extend({
     },
 
     markAllAsRead: function() {
+        this.stopListening(this.collection, 'change:unread');
```
Comment (alexgerstein) [inline sourcemash/frontend/static/js/views/items_view.js]: Interesting. Didn't know this existed.

File: sourcemash/frontend/static/js/views/items_view.js [hunk 6]
```diff
@@ -14,13 +15,21 @@ Sourcemash.Views.ItemsView = Backbone.View.extend({
...
+        this.model.save({unread: false});
```
Comment (alexgerstein) [inline sourcemash/frontend/static/js/views/items_view.js]: Do we need these two lines? If someone clicks "markAllAsRead" multiple times, it's still going to retrigger the api call because of the save.

File: sourcemash/frontend/static/js/views/items_view.js [hunk 7]
```diff
@@ -55,6 +64,9 @@ Sourcemash.Views.ItemsView = Backbone.View.extend({
         });
 
         this.itemViews = itemCards;
+        if (this.itemViews.length > 0) {
```
Comment (alexgerstein) [inline sourcemash/frontend/static/js/views/items_view.js]: What does this do?

File: sourcemash/frontend/templates/sidenav.tpl [hunk 8]
```diff
@@ -22,8 +22,8 @@
                 <li class="row collection-item">
                   <span class="col s7 truncate"><%= category.get('name') %></span>
                   <span class="badge col s2"><%=category.get('item_count')%></span>
-                  <% if (category.get('unread_count') > 0) { %>
-                    <span class="new badge col s3"><%=category.get('unread_count')%></span>
+                  <% if (category.get('unread')) { %>
+                    <span class="new badge col s3"></span>
```
Comment (alexgerstein) [inline sourcemash/frontend/templates/sidenav.tpl]: Should we say "New Items" or something?

Comment (alexgerstein) [inline sourcemash/frontend/templates/sidenav.tpl]: Or maybe just have a blue dot?

File: tests/factories/user_category_factories.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+from sourcemash.models import UserCategory, Category
+from user_factories import UserFactory
+
+class UserCategoryFactory(SQLAlchemyModelFactory):
+    FACTORY_FOR = UserCategory
+    FACTORY_SESSION = db.session
+
+    user = factory.SubFactory(UserFactory)
+    category = Category("category")
```
Comment (alexgerstein) [inline tests/factories/user_category_factories.py]: Should this be a subfactory as well?

File: sourcemash/api/categories.py (hunk 2 above)
Comment (alexgerstein) [inline sourcemash/api/categories.py]: I think we should cut it. It just adds time to API calls without serving enough functionality. It will also make the code much neater.

File: sourcemash/frontend/static/js/views/items_view.js (hunk 7 above)
Comment (swglad) [inline sourcemash/frontend/static/js/views/items_view.js]: Makes sure you only update the read status if the collection is nonempty. The problem is that this.collection isn’t always full when rendered gets called, and an empty collection has no notion of read status.

File: sourcemash/frontend/templates/sidenav.tpl (hunk 8 above)
Comment (swglad) [inline sourcemash/frontend/templates/sidenav.tpl]: It says “New”

**Comment by alexgerstein** [comment]
Freaking awesome work! I made a few comments on ways we can clean-up the code a little, but look tremendous overall.

File: sourcemash/api/categories.py (hunk 2 above)
Comment (swglad) [inline sourcemash/api/categories.py]: Done.

File: sourcemash/frontend/static/js/views/items_view.js (hunk 6 above)
Comment (swglad) [inline sourcemash/frontend/static/js/views/items_view.js]: We need to save since we need the database to sync against the UserFeed/UserCategory table now. Maybe the better way to do it is disable the button or make it only listenOnce to the event (Backbone has that built in)

Comment (alexgerstein) [inline sourcemash/frontend/static/js/views/items_view.js]: Won't the updateReadStatus get triggered from the for-loop where you do the save on the unread items?

File: sourcemash/frontend/static/js/views/items_view.js (hunk 7 above)
Comment (alexgerstein) [inline sourcemash/frontend/static/js/views/items_view.js]: If we only update change "unread" when an item is read in the collection, then do we need to also do this in the render at all?

File: sourcemash/frontend/static/js/views/items_view.js (hunk 6 above)
Comment (swglad) [inline sourcemash/frontend/static/js/views/items_view.js]: Yes, but there's no point for it to get hit 100 times for every item in a feed. Instead, I have it stopListening and just call it manually here (which is better I think).

File: sourcemash/frontend/static/js/views/items_view.js (hunk 7 above)
Comment (swglad) [inline sourcemash/frontend/static/js/views/items_view.js]: Yes, because now that there's no item_counts, you can't live update the notion of "New" for a category when clicking on items in a feed or for a feed when clicking on items in a category. This makes it such that when you open a new itemsView, it checks if all the items are read and updates it accordingly.

File: tests/factories/user_category_factories.py (hunk 9 above)
Comment (swglad) [inline tests/factories/user_category_factories.py]: We don't have a category factory and there isn't really a use for one (since we have itemWithCategory), so it just makes a Category here for this.

Comment (swglad) [inline tests/factories/user_category_factories.py]: But it does not make sense for the UserCategory factory to make an itemWithCategory just to get a category.

File: sourcemash/frontend/templates/feed.tpl [hunk 10]
```diff
@@ -1,6 +1,7 @@
 <div class="container">
   <div class="row valign-wrapper">
-      <h2 class="col s8 m9"><%= model.get('title') %></h2>
+      <h2 class="col s6 m9"><%= model.get('title') %></h2>
+      <span class="col s2 grey-text text-lighten-1">[<%=model.get('item_count')%> items]</span>
```
Comment (swglad) [inline sourcemash/frontend/templates/feed.tpl]: Let me know if you like this.

File: sourcemash/api/categories.py [hunk 11]
```diff
@@ -1,11 +1,23 @@
...
 from sqlalchemy import func
+from sqlalchemy.orm.exc import MultipleResultsFound, NoResultFound
 
+MIN_ITEM_COUNT = 1
+
+class isUnread(fields.Raw):
+    def output(self, key, category):
+        try:
+            unread = UserCategory.query.filter_by(user=current_user, category_id=category.id).one().unread
+        except (MultipleResultsFound, NoResultFound):
```
Comment (alexgerstein) [inline sourcemash/api/categories.py]: Should MultipleResultsFound ever happen?

File: sourcemash/api/categories.py [hunk 12]
```diff
@@ -13,18 +25,41 @@
...
+    def put(self, id):
+        args = self.reqparse.parse_args()
+
+        category = Category.query.get_or_404(id)
+
+        # Mark category as Read
+        if args.unread != None:
+            try:
+                user_category = UserCategory.query.filter_by(user=current_user, category_id=category.id).one()
+            except (MultipleResultsFound, NoResultFound):
```
Comment (alexgerstein) [inline sourcemash/api/categories.py]: Multiple?

File: sourcemash/api/categories.py [hunk 13]
```diff
@@ -44,24 +79,13 @@ def get(self):
...
-
-            category.item_count = count
-            category.unread_count = category.item_count - read_item_count
 
-            unsubscribed_item = Item.query.filter(Item.cats.contains(category)) \
+            if count > MIN_ITEM_COUNT:
+                unsubscribed_item = Item.query.filter(Item.cats.contains(category)) \
                                           .filter(~Item.feed_id.in_(user_feed_ids)) \
                                           .first()
-            if unsubscribed_item:
```
Comment (alexgerstein) [inline sourcemash/api/categories.py]: How are we handling unsubscribed items for the unread boolean?

File: sourcemash/api/categories.py [hunk 14]
```diff
@@ -44,24 +79,13 @@ def get(self):
...
-                                .filter(Item.cats.contains(category)) \
-                                .filter(~UserItem.unread) \
-                                .count()
-
-            category.item_count = count
-            category.unread_count = category.item_count - read_item_count
 
-            unsubscribed_item = Item.query.filter(Item.cats.contains(category)) \
+            if count > MIN_ITEM_COUNT:
+                unsubscribed_item = Item.query.filter(Item.cats.contains(category)) \
```
Comment (alexgerstein) [inline sourcemash/api/categories.py]: Do we even need this anymore?

File: sourcemash/api/feeds.py [hunk 15]
```diff
@@ -27,21 +28,17 @@ def output(self, key, feed):
...
-class getUnreadCount(fields.Raw):
+class isUnread(fields.Raw):
     def output(self, key, feed):
         if not current_user.is_authenticated():
-            return 0
+            return True
+
+        try:
+            unread = UserFeed.query.filter_by(user=current_user, feed_id=feed.id).one().unread
+        except (MultipleResultsFound, NoResultFound):
```
Comment (alexgerstein) [inline sourcemash/api/feeds.py]: Multiple?

File: sourcemash/api/feeds.py [hunk 16]
```diff
@@ -96,7 +93,7 @@ def post(self):
         # Get or Create Feed
         try:
             feed = Feed.query.filter(Feed.url==rss_feed['url']).one()
-        except:
+        except (MultipleResultsFound, NoResultFound):
```
Comment (alexgerstein) [inline sourcemash/api/feeds.py]: Multiple.

File: sourcemash/api/feeds.py [hunk 17]
```diff
@@ -120,7 +117,7 @@ def post(self):
         try:
             subscription = current_user.subscribed.filter(Feed.id==feed.id).one()
             return {"errors": {"url": ["Already subscribed"]}}, 409
-        except:
+        except (MultipleResultsFound, NoResultFound):
```
Comment (alexgerstein) [inline sourcemash/api/feeds.py]: Multiple?

File: sourcemash/frontend/static/js/views/items_view.js [hunk 18]
```diff
@@ -55,6 +66,7 @@ Sourcemash.Views.ItemsView = Backbone.View.extend({
         });
 
         this.itemViews = itemCards;
+        this.updateReadStatus();
```
Comment (alexgerstein) [inline sourcemash/frontend/static/js/views/items_view.js]: Why is this here again?

File: sourcemash/frontend/templates/feed.tpl (hunk 10 above)
Comment (alexgerstein) [inline sourcemash/frontend/templates/feed.tpl]: If this is the only time we use item_count, I'm not so sure it's worth it.

File: tests/factories/user_category_factories.py (hunk 9 above)
Comment (alexgerstein) [inline tests/factories/user_category_factories.py]: What's weird about categories made this way in the factory is that they won't actually be present in the database i don't think. So if you wanted to access category in a test, you'll get an error saying it's not bound to the session or something like that

File: sourcemash/api/categories.py (hunk 11 above)
Comment (swglad) [inline sourcemash/api/categories.py]: No, it shouldn't. But if there is a db problem and it happens, we should catch it and return the "default" behavior as opposed to dumping a stacktrace.

Comment (swglad) [inline sourcemash/api/categories.py]: Nope. Will remove.

File: sourcemash/api/categories.py (hunk 14 above)
Comment (swglad) [inline sourcemash/api/categories.py]: I was thinking about it, but I still like the feature. I think ideally it should be a setting that users turn on & off (just a column on the user table), but it's worth keeping.

File: sourcemash/frontend/static/js/views/items_view.js (hunk 18 above)
Comment (swglad) [inline sourcemash/frontend/static/js/views/items_view.js]: Let's say you read every item in every feed. Each category will still be marked as "New" in the sidebar since we're not tracking read/unread by counting items anymore. So, this makes sure that when you click on a Category that says "New" with all read items in it, it updates the read status and marks it as read. 

We need this because, otherwise, there is no change on "unread" so the category will continue to be marked as unread even if every item is read since there's no change.

File: sourcemash/frontend/templates/feed.tpl (hunk 10 above)
Comment (swglad) [inline sourcemash/frontend/templates/feed.tpl]: It still seems weird not to list the number of items anywhere. We can get away with it at the category level since there are never that many items in a category, but feeds range from 10 to 90 items so it's just kinda odd not to report that.

File: tests/factories/user_category_factories.py (hunk 9 above)
Comment (swglad) [inline tests/factories/user_category_factories.py]: Not an issue with tests and seems to work as expected...

File: sourcemash/api/categories.py (hunk 13 above)
Comment (swglad) [inline sourcemash/api/categories.py]: Not a problem since the PUT is triggered from the frontend so it just sees it as another item in the collection.

File: worker/scraper.py [hunk 19]
```diff
@@ -105,6 +109,11 @@ def _store_items(feed):
         db.session.add(new_entry)
         db.session.commit()
 
+        new_item_count += 1
+
+    feed.item_count += new_item_count
+    db.session.commit()
+
```
Comment (alexgerstein) [inline worker/scraper.py]: Need to retrigger unread = True after scraping.

Comment (swglad) [inline worker/scraper.py]: Merge in your scraper split first and put UserFeed at the end of feeds and UserCategory at the end of category

File: sourcemash/frontend/static/js/views/items_view.js [hunk 20]
```diff
@@ -3,6 +3,7 @@ Sourcemash.Views.ItemsView = Backbone.View.extend({
         if (this.model) {
             this.listenTo(this.model, 'change:subscribed change:title', this.render);
         };
+        this.listenTo(this.collection, 'change:unread', this.updateReadStatus);
```
Comment (alexgerstein) [inline sourcemash/frontend/static/js/views/items_view.js]: Try adding sync here.

Comment (swglad) [inline sourcemash/frontend/static/js/views/items_view.js]: Worked perfectly. Great call.

File: worker/scraper.py (hunk 19 above)
Comment (swglad) [inline worker/scraper.py]: Done.

File: worker/scraper.py [hunk 21]
```diff
@@ -20,6 +21,11 @@
 def scrape_feed_articles(feed):
     _store_items(feed)
 
+    # Mark feed as unread for all users
+    for user_feed in UserFeed.query.filter_by(feed_id=feed.id).all():
```
Comment (alexgerstein) [inline worker/scraper.py]: What if no new items came in on the _store_items call for the feed?

File: worker/scraper.py [hunk 22]
```diff
@@ -49,6 +55,15 @@ def categorize_feed_articles(feed, categorizer):
         for category in categories:
             item.categories.append(category)
 
+            # Mark category as unread for all users
+            try:
+                category_model = Category.query.filter_by(category=category).one()
+                for user_category in UserCategory.query.filter_by(category=category_model).all():
```
Comment (alexgerstein) [inline worker/scraper.py]: Just learned about: http://docs.sqlalchemy.org/en/latest/core/dml.html

Want to give that a shot? Looks nifty!

File: worker/scraper.py (hunk 21 above)
Comment (swglad) [inline worker/scraper.py]: Great catch. I'll have store items return a book new_items

File: worker/scraper.py (hunk 22 above)
Comment (swglad) [inline worker/scraper.py]: Not working for now but let's try it again another time.

File: worker/scraper.py (hunk 19 above)
Comment (swglad) [inline worker/scraper.py]: Actually, see directly below.

File: sourcemash/api/feeds.py [hunk 23]
```diff
@@ -55,8 +52,7 @@ def output(self, key, feed):
 }
 
 feed_status_fields = {
-    'item_count': getItemCount,
-    'unread_count': getUnreadCount
+    'unread': isUnread
```
Comment (alexgerstein) [inline sourcemash/api/feeds.py]: Move this into feed_fields too?

File: sourcemash/frontend/static/js/views/items_view.js [hunk 24]
```diff
@@ -15,12 +16,24 @@ Sourcemash.Views.ItemsView = Backbone.View.extend({
...
+            mixpanel.track("Marked all as read");
+            unread.forEach(function(model) {
+                model.save({unread: false});
+                mixpanel.people.increment("items read");
+            });
+        };
+    },
+
+    updateReadStatus: function() {
+        if (this.collection.length > 0) {
```
Comment (alexgerstein) [inline sourcemash/frontend/static/js/views/items_view.js]: Do we need this if check if it only happens after the feed is fetched/synced?

File: sourcemash/api/feeds.py (hunk 23 above)
Comment (swglad) [inline sourcemash/api/feeds.py]: Kk

File: sourcemash/frontend/templates/sidenav.tpl [hunk 25]
```diff
@@ -41,10 +38,9 @@
           <div class="loading center-align"></div>
           <% feeds.where({subscribed: true}).forEach(function(feed) { %>
             <a href="/#feeds/<%= feed.get('id') %>"><li class="row collection-item">
-            <span class="col s7 truncate"><%= feed.get('title') %></span>
-            <span class="badge col s2"><%=feed.get('item_count')%></span>
-            <% if (feed.get('unread_count') > 0) { %>
-              <span class="new badge col s3"><%=feed.get('unread_count')%></span>
+            <span class="col s9 truncate"><%= feed.get('title') %></span>
+            <% if (feed.get('unread')) { %>
+              <span class="new badge col s3"></span>
```
Comment (alexgerstein) [inline sourcemash/frontend/templates/sidenav.tpl]: Instead of the heavy new label, what if we have a dot made with icon: mdi-image-brightness-1. We can color it whatever fits our theme and give it a tool tip that says "New Items"? "new" probably isn't descriptive enough because a user wouldn't know if that means it has new items or if it's a new category.

File: tests/factories/user_category_factories.py (hunk 9 above)
Comment (alexgerstein) [inline tests/factories/user_category_factories.py]: It may become an issue down the line, but we can leave it for now.

File: worker/scraper.py (hunk 22 above)
Comment (alexgerstein) [inline worker/scraper.py]: Got it. It's even simpler than we thought:

```python
UserCategory.query.filter_by(category=category_model).update({"unread": True})
db.session.commit()
```

File: worker/scraper.py (hunk 19 above)
Comment (alexgerstein) [inline worker/scraper.py]:
```python
UserFeed.query.filter_by(feed_id=feed.id).update({"unread": True})
db.session.commit()
```

Comment (swglad) [inline worker/scraper.py]: Done. See below.

File: sourcemash/frontend/static/scss/sourcemash.scss [hunk 26]
```diff
@@ -184,6 +184,12 @@ div.side-nav {
   margin: 0;
 }
 
+.new-items-badge {
+  i {
+    color: #D32F2F;
```
Comment (alexgerstein) [inline sourcemash/frontend/static/scss/sourcemash.scss]: Can we use the @extend functionality here?
